scapy/contrib/lldp.py

```diff
@@ -0,0 +1,714 @@
+
+        This module is free software; you can redistribute it and/or
+        modify it under the terms of the GNU General Public License
+        as published by the Free Software Foundation; either version 2
+        of the License, or (at your option) any later version.
+
+        This module is distributed in the hope that it will be useful,
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU General Public License for more details.
+
+    :description:
+
+        This module provides Scapy layers for the LLDP protocol.
+
+        normative references:
+            - IEEE 802.1AB 2016 - LLDP protocol, topology and MIB description
+
+    :TODO:
+        - organization specific TLV e.g. ProfiNet
+
+    :NOTES:
+        - you can find the layer configuration options at the end of this file
+        - default configuration enforces standard conform
+          - frame structure
+                (ChassisIDTLV/PortIDTLV/TimeToLiveTLV/.../EndofLLDPDUTLV)
+          - multiplicity of TLVs (if given by the standard)
+          - min sizes of strings used by the TLVs
+        - conf.contribs['LLDP'].strict_mode_disable() -> disable strict mode
+        - strict mode = True => conf.debug_dissector = True
+
+"""
+from scapy.config import conf
+from scapy.layers.dot11 import Packet
+from scapy.layers.l2 import Ether, Dot1Q, bind_layers, \
+    struct, BitField, StrLenField, ByteEnumField, BitEnumField, \
+    BitFieldLenField, ShortField, Padding, Scapy_Exception, \
+    XStrLenField
+
+LLDP_NEAREST_BRIDGE_MAC = '01:80:c2:00:00:0e'
+LLDP_NEAREST_NON_TPMR_BRIDGE_MAC = '01:80:c2:00:00:03'
+LLDP_NEAREST_CUSTOMER_BRIDGE_MAC = '01:80:c2:00:00:00'
+
+LLDP_ETHER_TYPE = 0x88cc
+
+
+class LLDPInvalidFrameStructure(Scapy_Exception):
+    """
+    basic frame structure not standard conform
+    (missing TLV, invalid order or multiplicity)
+    """
+    pass
+
+
+class LLDPInvalidLastLayerException(Scapy_Exception):
+    """
+    in strict mode, last layer in frame must be of type LLDPDUEndOfLLDPDU
+    """
+    pass
+
+
+class LLDPMissingLowerLayer(Scapy_Exception):
+    """
+    first layer below first LLDPDU must be Ethernet or Dot1q
+    """
+    pass
+
+
+class LLDPInvalidTLVCount(Scapy_Exception):
+    """
+    invalid number of entries for a specific TLV type
+    """
+    pass
+
+
+class LLDPInvalidLengthField(Scapy_Exception):
+    """
+    invalid value of length field
+    """
+    pass
+
+
+class LLDPDU(Packet):
+    """
+    base class for all LLDP data units
+    """
+    TYPES = {
+        0x00: 'end of LLDPDU',
+        0x01: 'chassis id',
+        0x02: 'port id',
+        0x03: 'time to live',
+        0x04: 'port description',
+        0x05: 'system name',
+        0x06: 'system description',
+        0x07: 'system capabilities',
+        0x08: 'management address',
+        xrange(0x09, 0x7e): 'reserved - future standardization',
```

Comment:
As specified in CONTRIBUTING.md, please use `range` from `scapy.modules.six.moves`